### PR TITLE
Branch for bank service practice03

### PR DESCRIPTION
**Pull request by AntonPanchishko**

### CODE REVIEW

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 1]
```diff
@@ -45,6 +48,11 @@ public void transfer(Account fromAccount, Account toAccount, int amount) {
         transactionIncoming.setAmount(BigDecimal.valueOf(amount));
         transactionIncoming.setDate(LocalDateTime.now());
         transactionIncoming.setType(Transaction.Type.INCOMING);
+        if (fromAccount.getCurrency() != toAccount.getCurrency()) {
+            double rate = clientService.getRate(LocalDate.now(),
+                    fromAccount.getCurrency(), toAccount.getCurrency());
+            amount = (int) (amount * rate);
```
Comment (boroda4436) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java]: better to have amount as a BigDecimal type

File: src/main/java/com/example/bankservicedemo/service/ClientService.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+public class ClientService {
+    @Value(value = "$(service.url)")
+    private String url;
+
+    public double getRate(LocalDate date, Currency currencyFrom, Currency currencyTo) {
+        String path = url + currencyFrom.toString()
+                + "&to=" + currencyTo.toString() + "&date=" + date;
+        try {
+            URL url = new URL(path);
+            HttpURLConnection request = (HttpURLConnection) url.openConnection();
```
Comment (boroda4436) [inline src/main/java/com/example/bankservicedemo/service/ClientService.java]: Better move it to a HttpClient component

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bankservicedemo/service/ClientService.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    private final HttpClient httpClient;
+
+    @Autowired
+    public ClientService(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public double getRate(LocalDate date, Currency currencyFrom, Currency currencyTo) {
+        String path = url + currencyFrom.toString()
+                + "&to=" + currencyTo.toString() + "&date=" + date;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/ClientService.java:22]: it is better to use String.format() in such cases

File: src/main/java/com/example/bankservicedemo/service/HttpClient.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        JsonObject jsonObject = null;
+        try {
+            URL url = new URL(path);
+            HttpURLConnection request = (HttpURLConnection) url.openConnection();
+            request.connect();
+            JsonElement root = JsonParser
+                    .parseReader(new InputStreamReader((InputStream) request.getContent()));
+            jsonObject = root.getAsJsonObject();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/HttpClient.java:25]: better throw RuntimeException

File: src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java [hunk 5]
```diff
@@ -20,33 +22,39 @@
...
         transactionIncoming.setAccountFrom(fromAccount);
         transactionIncoming.setAccountTo(toAccount);
-        transactionIncoming.setAmount(BigDecimal.valueOf(amount));
+        transactionIncoming.setAmount(amount);
         transactionIncoming.setDate(LocalDateTime.now());
         transactionIncoming.setType(Transaction.Type.INCOMING);
+        if (fromAccount.getCurrency() != toAccount.getCurrency()) {
+            double rate = clientService.getRate(LocalDate.now(),
+                    fromAccount.getCurrency(), toAccount.getCurrency());
+            amount = BigDecimal.valueOf(Double.parseDouble(String.valueOf(amount)) * rate);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bankservicedemo/service/impl/TransactionServiceImpl.java:54]: or we could use `multiply()` method on 2 BigDecimals

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**